### PR TITLE
bindings: allow use of `pathlib.Path | str` for paths

### PR DESCRIPTION
**Pull request by nim65s**
Hi,

This update the python bindings to allow passing filenames through [`pathlib`](https://docs.python.org/3/library/pathlib.html) objects, without breaking compatibility with the previous way of passing filenames through bare `str`.

It also modernize python code by adding the `PTH` ruff check and update `os.path` code with `pathlib`.

### CODE REVIEW

File: bindings/python/utils/path.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
+//
+// Copyright (c) 2024 CNRS
+//
+
+// #include <string>
+
+// #include <boost/python.hpp>
```
Comment (jcarpent) [inline bindings/python/utils/path.cpp:7]: Could you remove these lines @nim65s?

Comment (nim65s) [reply bindings/python/utils/path.cpp]: oops, done

**Comment by nim65s** [comment]
windows CI still failing with unrelated:
```
C:\Miniconda\envs\pinocchio\Library\include\cppad/cg/cg.hpp(200): error C2063: 'CppAD::cg::operator *': not a function
C:\Miniconda\envs\pinocchio\Library\include\cppad/cg/base_double.hpp(27): note: see reference to class template instantiation 'CppAD::cg::CG<double>' being compiled
C:\Miniconda\envs\pinocchio\Library\include\cppad/cg/cg.hpp(217): error C2143: syntax error: missing ';' before '<'
C:\Miniconda\envs\pinocchio\Library\include\cppad/cg/cg.hpp(217): error C2433: 'CppAD::cg::NearEqual': 'friend' not permitted on data declarations
C:\Miniconda\envs\pinocchio\Library\include\cppad/cg/cg.hpp(217): error C2365: 'CppAD::cg::NearEqual': redefinition; previous definition was 'function'
```

**Review by jcarpent — APPROVED** [review]

**Comment by jorisv** [comment]
`build_all` label is adding -DBUILD_WITH_CODEGEN_SUPPORT=ON on Windows but this option is not allowed on this OS.

We will fix this.

**Review by jorisv — DISMISSED** [review]

**Review by jcarpent — APPROVED** [review]

**Pull request merged**